### ISSUE TITLE
加载Randeng-T5-784M继续预训练，开启FP16后loss nan

### ISSUE THREAD

**Issue by miaodog** [IDEA-CCNL/Fengshenbang-LM#222]
显卡：A100（80G）

相关依赖：
pip install pytorch-lightning==1.7.7
pip install transformers==4.15.0
pip install deepspeed==0.7.4
pip install sentencepiece==0.1.97

预训练模型：Randeng-T5-784M
语料：自有语料，大概量级3000W

模型代码：
output = self.model(input_ids=batch['input_ids'], labels=batch['labels'])
self.log('train_loss', output.loss, on_step=True, on_epoch=True, prog_bar=True, sync_dist=True)

optimizer = Adafactor(optimizer_grouped_parameters, scale_parameter=False, relative_step=False, warmup_init=False, lr=self.hparams.learning_rate)
scheduler = get_linear_schedule_with_warmup(
    optimizer,
    num_warmup_steps=num_warmup_steps,
    num_training_steps=self.total_steps,
)

训练参数：
    --precision 16 \
    --num_nodes 1 \
    --gpus 8 \
    --strategy ddp \
    --max_epochs 1 \
    --learning_rate 1e-4 \
    --weight_decay 0.01 \
    --adam_beta1 0.9 \
    --adam_beta2 0.999 \
    --adam_epsilon 1e-6 \

现象：
precision=16，loss nan（如下所示），precision=32或者bf16则正常

Epoch 0/1 | Batch 0/15582 | Train Loss nan | Learning Rate 0.0
Epoch 0/1 | Batch 100/15582 | Train Loss nan | Learning Rate 2e-08
Epoch 0/1 | Batch 200/15582 | Train Loss nan | Learning Rate 4e-08
Epoch 0/1 | Batch 300/15582 | Train Loss nan | Learning Rate 6e-08

**Comment by GGGGGGXY** [comment]
抛开数据问题的话，一般loss为nan可能是：
1.学习率过大
2.新初始化的层参数量级不对
3.模型精度不够

这里如果fp32和bf16没问题的话，应该是精度问题了，两个尝试的方向
1.一个就直接用bf16开始训练 
2. initial_scale_power 设置大一点，比如32、64，看看会不会再出现nan的问题。

**Comment by GGGGGGXY** [comment]
https://github.com/IDEA-CCNL/Fengshenbang-LM/tree/main/fengshen/examples/qa_t5

这个是我们在rangdeng-t5-784M上面的一个finetune示例，也可以参考一下

**Comment by miaodog** [comment]
> initial_scale_power

1. 想训练速度再快一些，所以最好能使用FP16；
2. initial_scale_power 这个是deepspeed的参数么？如果我使用ddp，该怎么办呢？

**Comment by GGGGGGXY** [comment]
1.如果你用的Ampere架构的显卡，bf16也是有加速的
2.我以为你用的是deepspeed，如果没有用的话就木有这个参数的。


这里优化器有尝试过使用AdamW试试吗

**Comment by GGGGGGXY** [comment]
> > initial_scale_power
> 
> 1. 想训练速度再快一些，所以最好能使用FP16；
> 2. initial_scale_power 这个是deepspeed的参数么？如果我使用ddp，该怎么办呢？

我看到你们显卡是A100，这个对bf16是有优化的，另外8张卡推荐可以试试deepspeed，加速效果很明显

**Comment by GGGGGGXY** [comment]
> > initial_scale_power
> 
> 1. 想训练速度再快一些，所以最好能使用FP16；
> 2. initial_scale_power 这个是deepspeed的参数么？如果我使用ddp，该怎么办呢？

我粗略的了扫一了这个issue，感觉跟你遇到的问题比较类似
https://github.com/huggingface/transformers/pull/10956

可以看看有启发不

**Comment by miaodog** [comment]
> huggingface/transformers#10956

嗯嗯，好的，我再测试下deepspeed和AdamW优化器看看

**Comment by miaodog** [comment]
> > > initial_scale_power
> > 
> > 
> > 
> > 1. 想训练速度再快一些，所以最好能使用FP16；
> > 2. initial_scale_power 这个是deepspeed的参数么？如果我使用ddp，该怎么办呢？
> 
> 我粗略的了扫一了这个issue，感觉跟你遇到的问题比较类似 [huggingface/transformers#10956](https://github.com/huggingface/transformers/pull/10956)
> 
> 可以看看有启发不

这个我也check下
